### PR TITLE
Provide a custom Asciidoctor include macro

### PR DESCRIPTION
**Pull request by meistermeier**
uses a ruby based solution to add snippet blocks to the parent block element

Fixes gh-323

### CODE REVIEW

File: spring-restdocs-asciidoctor/src/main/java/org/springframework/restdocs/asciidoctor/RestDocsExtensionRegistry.java [hunk 1]
```diff
@@ -28,8 +30,15 @@
 
 	@Override
 	public void register(Asciidoctor asciidoctor) {
-		asciidoctor.javaExtensionRegistry()
-				.preprocessor(new DefaultAttributesPreprocessor());
+		final JavaExtensionRegistry registry = asciidoctor.javaExtensionRegistry();
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/main/java/org/springframework/restdocs/asciidoctor/RestDocsExtensionRegistry.java]: REST Docs' code style is to only use `final` when it's required

File: spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
+require 'asciidoctor/extensions'
+require 'stringio'
+
+# Spring REST Docs block macro to import multiple snippet of an operation at once
+#
+# Usage
+#
+#   srd::operation-name[snippets='snippet-name1,snippet-name2', level=<indentation level>]
+#
+class RestDocsBlockMacro < Asciidoctor::Extensions::BlockMacroProcessor
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb]: I think this should be called `OperationBlockMacro` and the file itself should be called `operation_block_macro.rb`

File: spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+               :'response-fields' => 'Response fields',
+               :'links' => 'Links'
+    }
+  end
+
+  def process(parent, reader, attrs)
+    doc = parent.document
+    snippet_dir = doc.attributes['snippets']
+    snippets = snippets_from_attributes(attrs)
+    section_level = section_level_from_attributes(attrs)
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb]: I think we could use `parent.level` to figure this out automatically

File: spring-restdocs-asciidoctor/src/main/java/org/springframework/restdocs/asciidoctor/RestDocsExtensionRegistry.java [hunk 4]
```diff
@@ -28,8 +30,15 @@
...
+		registry.preprocessor(new DefaultAttributesPreprocessor());
+
+		RubyExtensionRegistry rubyExtensionRegistry = asciidoctor.rubyExtensionRegistry();
+		rubyExtensionRegistry
+				.loadClass(RestDocsExtensionRegistry.class
+						.getResourceAsStream("/extensions/srd-include-block-macro.rb"))
+				.blockMacro("srd", "RestDocsBlockMacro");
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/main/java/org/springframework/restdocs/asciidoctor/RestDocsExtensionRegistry.java]: I don't think the macro should be called `srd` as that would seem to imply the REST Docs will only ever have a single block macro. There doesn't seem to be a tradition of trying to namespace macro names in Asciidoctor so I'm tempted to go with `operation` instead.

File: spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+				"</div>";
+
+		String result = this.asciidoctor.convert(
+				"srd::some-operation[snippets='curl-request']", this.options);
+
+		assertThat(result, equalTo(expectedContent));
+	}
+
+	@Test
+	public void useMacroWithoutSnippetAddsNothing() {
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java]: I'm intrigued by this test. What was the reasoning behind adding nothing? I would expect this to either produce an error or to include all of the snippets for a particular operation. My preference is for the latter but I'm not set on that.

File: spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    end
+  end
+
+  # if no section indentation level is provided, default to 4
+  def section_level_from_attributes(attrs)
+    (attrs['level'].to_i if attrs['level']) || 4
+  end
+
+  def snippets_from_attributes(attrs)
+    (attrs['snippets'].split(',') if attrs['snippets']) || []
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb]: I think using [`fetch`](https://ruby-doc.org/core-1.9.3/Hash.html#method-i-fetch) would make this read a bit better:

```ruby
attrs.fetch('snippets', '').split(',')
```

File: spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  end
+
+  def append_snippet_block(content, snippet, params)
+    write_title content, snippet, params[:section_level]
+    write_content content, snippet, params
+  end
+
+  def write_content(content, snippet, params)
+    asciidoc_extension = '.adoc'
+    snippet_path = params[:snippet_dir].to_s + File::SEPARATOR +
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb]: I think using `File.join` and some string interpolation would improve the readability here. Something like:

```ruby
File.join(params[:snippet_dir], params[:operation], "#{snippet}.adoc")
```

File: spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        params[:operation] + File::SEPARATOR +
+        snippet + asciidoc_extension
+
+    content.puts File.readlines(snippet_path).join
+
+  rescue Errno::ENOENT
+    content.puts "WARNING: snippet not found: " + snippet_path
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb]: I'd use string interpolation here too

File: spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    add_new_line content
+  end
+
+  def write_title(content, snippet, level)
+    section_level = '=' * level
+    content.puts section_level + ' ' + title(snippet)
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/main/resources/extensions/srd-include-block-macro.rb]: I'd use string interpolation here too

File: spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for ruby based rest docs block macro
+ * Because there is no java implementation (yet)
+ * we can only test the behaviour when rendering
+ *
+ * @author Gerrit Meier
+ */
+public class RestDocsIncludeBlockMacroTest {
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java]: This should be named `…Tests` to match the other tests classes

File: spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+	private Attributes getAttributes() {
+		Attributes attributes = new Attributes();
+		attributes.setAttribute("projectdir", new File(".").getAbsolutePath());
+		return attributes;
+	}
+
+	@Test
+	public void includeSnippetWithLevel() {
+		String expectedContent = "<div class=\"sect1\">\n" +
```
Comment (wilkinsona) [inline spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java]: My preference would be to store the expected content in files in `src/test/resources`

**Comment by meistermeier** [comment]
Thanks for the feedback. I think I can do the changes but this may take until the next weekend.

File: spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java (hunk 5 above)
Comment (meistermeier) [inline spring-restdocs-asciidoctor/src/test/java/org/springframework/restdocs/asciidoctor/RestDocsIncludeBlockMacroTest.java]: Now it adds every exiting snippet in the operation folder.

**Comment by meistermeier** [comment]
I have updated the code and rebased it (so that the code comments above are broken sorry for that 🙂  ). 
Major changes:
* the include without any snippets will include every snippet in the operation directory
* the level of the include will be derived from the parent block level

**Comment by wilkinsona** [comment]
Thanks, @meistermeier. Did you miss some files when you made the update? The new tests are failing for me because they can't find the `build/generated-snippets/some-operation` directory

**Comment by meistermeier** [comment]
There went something completely wrong with the update...so I went down in the code dungeon and found some commits with loose endings ;)

**Comment by wilkinsona** [comment]
@meistermeier I've merged this into master. Thank you very much for the contribution.

@mojavelinux We couldn't have done this without your help. Thank you very much.

**Pull request closed without merging**